### PR TITLE
[sdformat13] new port

### PR DESCRIPTION
**Pull request by talregev**
Add sdformat13 port:

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by talregev** [comment]
I got this error:
```
CMake Error: install(EXPORT "sdformat13" ...) includes target "sdformat13" which requires target "using_parser_urdf" that is not in any export set.
CMake Error in src/CMakeLists.txt:
  export called with target "sdformat13" which requires target
  "using_parser_urdf" that is not in any export set.
```

**Comment by talregev** [comment]
logs:
[sdformat13.zip](https://github.com/microsoft/vcpkg/files/11410407/sdformat13.zip)

**Comment by Adela0814** [comment]
Note: I will be converting your PR to draft status. If you fixed the issue with CI, please reactivate it.

**Comment by talregev** [comment]
@Adela0814 Can you explain why this error is happening? Is it come from vcpkg? 
I try to look on it on the internet, and I didn't any export on the sdformat repo cod side. 
vcpkg make export?

**Comment by talregev** [comment]
@autoantwort @dg0yt @@traversaro
Can you help?

**Comment by dg0yt** [comment]
Grep for `using_parser_urdf` in the source code and see if it either can be added to the export set or recconstructed in the installed config file?

**Comment by traversaro** [comment]
I think it is a bug upstream in vcpkg in the case `BUILD_SHARED_LIBS` is set to `OFF`. In https://github.com/gazebosim/sdformat/blob/38f6435e095b5d82984aa26bbeac40ce08e34229/src/CMakeLists.txt#L118 the `sdformat13` links as `PRIVATE` the `using_parser_urdf` INTERFACE target. As long as  `BUILD_SHARED_LIBS` is set to `ON`, everything works fine as private linked libraries do not need to be found when `find_package(sdformat13)` is called, but if `BUILD_SHARED_LIBS` is set to `OFF`, then `using_parser_urdf` needs to be found when `find_package(sdformat13)` is called, and hence it should be installed. I guess the error went unnoticed upstream as there are not  a lot of users for the case when `BUILD_SHARED_LIBS` is set to `OFF`.

**Comment by talregev** [comment]
> Grep for `using_parser_urdf` in the source code and see if it either can be added to the export set or recconstructed in the installed config file?

It added on the cmake:
https://github.com/gazebosim/sdformat/blob/sdformat13_13.0.0/src/CMakeLists.txt#L26

**Comment by talregev** [comment]
> I think it is a bug upstream in vcpkg in the case `BUILD_SHARED_LIBS` is set to `OFF`

It by design. Can we compile it only with static?

**Comment by dg0yt** [comment]
> > I think it is a bug upstream in vcpkg in the case `BUILD_SHARED_LIBS` is set to `OFF`

To be clear, AFAIU it is a bug in upstream which becomes visible in vcpkg's static triplets.

> It by design. Can we compile it only with static? Can we patch it, or make gazebo upstream changes?

I don't see how this is "by design". It should be fixed upstream, and the same patch can be applied in the vcpkg port for the current release.

**Comment by talregev** [comment]
> > > I think it is a bug upstream in vcpkg in the case `BUILD_SHARED_LIBS` is set to `OFF`
> 
> To be clear, AFAIU it is a bug in upstream which becomes visible in vcpkg's static triplets.
> 
> > It by design. Can we compile it only with static? Can we patch it, or make gazebo upstream changes?
> 
> I don't see how this is "by design". It should be fixed upstream, and the same patch can be applied in the vcpkg port for the current release.

I meant it by design that vcpkg turn off the shared libs in Linux compilation.

**Comment by dg0yt** [comment]
> > I don't see how this is "by design". It should be fixed upstream, and the same patch can be applied in the vcpkg port for the current release.
> 
> I meant it by design that vcpkg turn off the shared libs in Linux compilation.

Oh, I don't think this is by design, but just the historic choice of the "supported triplet" `x64-linux`. There is also a `x64-linux-dynamic` "community triplet".

**Comment by talregev** [comment]
@traversaro But it also fail on windows that have shared libs on. Why?

**Comment by dg0yt** [comment]
~~~
Make Error at D:/installed/x64-windows/share/cmake/gz-cmake3/cmake3/GzConfigureBuild.cmake:73 (message):
  -- BUILD ERRORS: These must be resolved before compiling.
Call Stack (most recent call first):
  CMakeLists.txt:157 (gz_configure_build)


CMake Error at D:/installed/x64-windows/share/cmake/gz-cmake3/cmake3/GzConfigureBuild.cmake:75 (message):
  -- 	Couldn't find the urdfdom >= 1.0 system installation
Call Stack (most recent call first):
  CMakeLists.txt:157 (gz_configure_build)


CMake Error at D:/installed/x64-windows/share/cmake/gz-cmake3/cmake3/GzConfigureBuild.cmake:77 (message):
  -- END BUILD ERRORS

Call Stack (most recent call first):
  CMakeLists.txt:157 (gz_configure_build)


CMake Error at D:/installed/x64-windows/share/cmake/gz-cmake3/cmake3/GzConfigureBuild.cmake:82 (message):
  Errors encountered in build.  Please see BUILD ERRORS above.
Call Stack (most recent call first):
  CMakeLists.txt:157 (gz_configure_build)
~~~

**Comment by dg0yt** [comment]
The new gz port are still not mature.

~~~
Make Error at D:/installed/x64-windows/share/cmake/gz-cmake3/cmake3/GzConfigureBuild.cmake:73 (message):
  -- BUILD ERRORS: These must be resolved before compiling.
Call Stack (most recent call first):
  CMakeLists.txt:157 (gz_configure_build)


CMake Error at D:/installed/x64-windows/share/cmake/gz-cmake3/cmake3/GzConfigureBuild.cmake:75 (message):
  -- 	Couldn't find the urdfdom >= 1.0 system installation
Call Stack (most recent call first):
  CMakeLists.txt:157 (gz_configure_build)


CMake Error at D:/installed/x64-windows/share/cmake/gz-cmake3/cmake3/GzConfigureBuild.cmake:77 (message):
  -- END BUILD ERRORS

Call Stack (most recent call first):
  CMakeLists.txt:157 (gz_configure_build)


CMake Error at D:/installed/x64-windows/share/cmake/gz-cmake3/cmake3/GzConfigureBuild.cmake:82 (message):
  Errors encountered in build.  Please see BUILD ERRORS above.
Call Stack (most recent call first):
  CMakeLists.txt:157 (gz_configure_build)
~~~

**Comment by talregev** [comment]
@dg0yt Can you fix it?

**Comment by dg0yt** [comment]
I guess for Windows, urdfdom needs https://github.com/ros/urdfdom/pull/169.

**Comment by talregev** [comment]
> I guess for Windows, urdfdom needs [ros/urdfdom#169](https://github.com/ros/urdfdom/pull/169).

ok. I will appreciate your fix.

**Comment by dg0yt** [comment]
>  ~~~
> -- Configuring done
> CMake Error: install(EXPORT "sdformat13" ...) includes target "sdformat13" which requires target "using_parser_urdf" that is not in any export set.
> CMake Error in src/CMakeLists.txt:
>   export called with target "sdformat13" which requires target
>   "using_parser_urdf" that is not in any export set.


The remaining build error is a bug in the project's files and unrelated to windows. 
The build system uses custom (gz) functions instead of pure CMake. 
Look/ask upstream.

**Comment by traversaro** [comment]
The error is the one described in https://github.com/microsoft/vcpkg/pull/31280#issuecomment-1537483083 , nothing related to the use of `gz_` CMake function.

**Comment by talregev** [comment]
> The error is the one described in [#31280 (comment)](https://github.com/microsoft/vcpkg/pull/31280#issuecomment-1537483083) , nothing related to the use of `gz_` CMake function.

Please create a PR in upstream, and I will make a patch from it.

**Comment by dg0yt** [comment]
> The error is the one described in [#31280 (comment)](https://github.com/microsoft/vcpkg/pull/31280#issuecomment-1537483083) , nothing related to the use of `gz_` CMake function.

`using_parser_urdf` is an interface target created in sdformat13 `src/CMakeLists.txt`, and linked to the main library target. This is the standard CMake stuff.
However, the export of CMake config is hidden in gz functions. This export would need to handle the interface target as well.

**Comment by talregev** [comment]
> > The error is the one described in [#31280 (comment)](https://github.com/microsoft/vcpkg/pull/31280#issuecomment-1537483083) , nothing related to the use of `gz_` CMake function.
> 
> `using_parser_urdf` is an interface target created in sdformat13 `src/CMakeLists.txt`, and linked to the main library target. This is the standard CMake stuff. However, the export of CMake config is hidden in gz functions. This export would need to handle the interface target as well.

Do you know how to solve it?
Can we make a patch for it?

**Comment by dg0yt** [comment]
> Do you know how to solve it?

That's the point I'm trying to make: I know how to use CMake. But once projects start to use their own language of macros and functions, it becomes tedious to find out where it breaks and how to fix it. We don't even have a CMake call stack to start from because the error shows too late.

**Comment by traversaro** [comment]
> > The error is the one described in [#31280 (comment)](https://github.com/microsoft/vcpkg/pull/31280#issuecomment-1537483083) , nothing related to the use of `gz_` CMake function.
> 
> `using_parser_urdf` is an interface target created in sdformat13 `src/CMakeLists.txt`, and linked to the main library target. This is the standard CMake stuff. However, the export of CMake config is hidden in gz functions. This export would need to handle the interface target as well.

Ahh, sorry I misunderstdood, that's totally correct. 

> @traversaro
> Please create a PR in upstream, and I will make a patch from it.

I am happy to give you the hints to understand how to fix it. To be honest I am not interest at the moment in building sdformat13 as part of vcpkg and in static mode, so I prefer not to prepare the PR and submit it.

The `sdformat13` library is created in https://github.com/gazebosim/gz-cmake/blob/gz-cmake3_3.1.0/cmake/GzUtils.cmake#L1540 and installed in https://github.com/gazebosim/gz-cmake/blob/gz-cmake3_3.1.0/cmake/GzUtils.cmake#L1611 , so the `EXPORT` used for it is `sdformat13` . The target installed with that export are installed in https://github.com/gazebosim/gz-cmake/blob/88e7dc5f727e2270f206f19f558ab8f7fb5b8df1/cmake/GzPackaging.cmake#L372 . So a simple workaround is to just install the `using_parser_urdf` with the `sdformat13` export, close to where `using_parser_urdf` library is defined with `add_library`.

**Comment by dg0yt** [comment]
Okay, I managed to add the install line. I will send a PR for this port to @talregev, adding a few more changes.

**Comment by talregev** [comment]
@traversaro Please stop ignore me at gazebo-classic repo, and review my PRs.
Thank you.

**Comment by dg0yt** [comment]
https://github.com/talregev/vcpkg/pull/1/files

**Comment by traversaro** [comment]
> @traversaro Please stop ignore me at gazebo-classic repo, and review my PRs. Thank you.

@traversaro I am not sure this is relevant here. More in general, do you realize that open-source mantaners you interact with with are complete strangers that do not owe you anything, and pinging them in unrelated issues you are not helping them finding the time to review PRs or work on features?

**Comment by talregev** [comment]
> > @traversaro Please stop ignore me at gazebo-classic repo, and review my PRs. Thank you.
> 
> I am not sure this is relevant here. More in general, do you realize that open-source mantaners you interact with with are complete strangers that do not owe you anything, and pinging them in unrelated issues you are not helping them finding the time to review PRs or work on features?

So I ping non relevant maintainers on my PRs. Is it a good reason to punish me and ignore me? 
I ping them for low respond on that repo. I wasn't sure how I can bring it up.

**Comment by talregev** [comment]
@traversaro We can continue that discussion on my PRs on gazebo classic repo.

**Comment by traversaro** [comment]
> I wasn't sure how I can bring it up.

(Sorry for the OT) My tipical strategy is to comment on the relevant PR (not somewhere else) "Can I do anything to help getting this PR?" (to sound a bit proactive), and avoid commenting elsewhere.

**Comment by talregev** [comment]
@dg0yt It not found urdfdom on windows:
[config-x86-windows-out.log.txt](https://github.com/microsoft/vcpkg/files/11470334/config-x86-windows-out.log.txt)

**Comment by dg0yt** [comment]
yes.

**Comment by talregev** [comment]
> yes.

How we can solve it?

**Comment by talregev** [comment]
@traversaro @dg0yt 
Any news regarding this PR?

**Comment by dg0yt** [comment]
Aren't you working on it?

**Comment by talregev** [comment]
> Aren't you working on it?

I don't know how to solve the problems. Should I close this PR?

**Comment by talregev** [comment]
@Adela0814 please review.

File: ports/sdformat13/cmake-config.patch [hunk 1]
```diff
@@ -0,0 +1,12 @@
+diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
+index 3465a8f..2b57440 100644
+--- a/src/CMakeLists.txt
++++ b/src/CMakeLists.txt
+@@ -26,6 +26,7 @@ else()
+   target_link_libraries(using_parser_urdf INTERFACE
+     GzURDFDOM::GzURDFDOM)
+ endif()
++install(TARGETS using_parser_urdf EXPORT ${PROJECT_LIBRARY_TARGET_NAME} COMPONENT libraries)
```
Comment (Adela0814) [inline ports/sdformat13/cmake-config.patch:9]: Please add `unofficial-` as a prefix.
see https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md#:~:text=To%20that%20end%2C%20any%20CMake%20configs%20that%20the%20port%20exports%2C%20which%20are%20not%20in%20the%20upstream%20library%2C%20should%20have%20unofficial%2D%20as%20a%20prefix.%20Any%20additional%20targets%20should%20be%20in%20the%20unofficial%3A%3A%3Cport%3E%3A%3A%20namespace.

File: ports/sdformat13/usage [hunk 2]
```diff
@@ -0,0 +1,4 @@
+sdformat13 provides CMake targets:
+
+    find_package(sdformat13 CONFIG REQUIRED)
+    target_link_libraries(main PRIVATE sdformat13::sdformat13)
```
Comment (Adela0814) [inline ports/sdformat13/usage:4]: There should be an automatically exported usage, you don't need to add it yourself.

File: ports/sdformat13/no-absolute.patch [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ 
+diff --git a/src/SDF.cc b/src/SDF.cc
+index 20dcd4c6..802cbde3 100644
+--- a/src/SDF.cc
++++ b/src/SDF.cc
+@@ -97,22 +97,6 @@ std::string findFile(const std::string &_filename, bool _searchLocalPath,
+     filename = filename.substr(idx + sep.length());
+   }
+ 
+-  // Next check the install path.
```
Comment (Adela0814) [inline ports/sdformat13/no-absolute.patch:19]: The changes of this patch should be submitted to the upstream, and we need to wait for the upstream response.

**Review by Adela0814 — CHANGES_REQUESTED** [review]

**Comment by Adela0814** [comment]
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". That way, I can be aware that you've responded since you can't modify the tags.

File: ports/sdformat13/no-absolute.patch (hunk 3 above)
Comment (talregev) [reply ports/sdformat13/no-absolute.patch:19]: This patch is an old patch, vcpkg have it from the beginning. 
https://github.com/microsoft/vcpkg/blob/master/ports/sdformat9/no-absolute.patch

File: ports/sdformat13/cmake-config.patch (hunk 1 above)
Comment (talregev) [reply ports/sdformat13/cmake-config.patch:9]: Can you write me the code I should write? I understand if it was `find_packge(...)`
But when it install, I am not sure where to put the `unofficial-` as a prefix.

I will glad for your help.

File: ports/sdformat13/usage (hunk 2 above)
Comment (talregev) [reply ports/sdformat13/usage]: done.

File: ports/sdformat13/no-absolute.patch (hunk 3 above)
Comment (talregev) [reply ports/sdformat13/no-absolute.patch:19]: @traversaro What do you say about this patch?

**Comment by Adela0814** [comment]
@talregev  I'm a bit confused, does `sdformat` require so many different versions of ports`(sdformat6, sdformat9, sdformat10, sdformat13)`? Instead of updating existing ports?

**Comment by talregev** [comment]
> @talregev I'm a bit confused, does `sdformat` require so many different versions of ports`(sdformat6, sdformat9, sdformat10, sdformat13)`? Instead of updating existing ports?

we already have this discussions on gz-ports:
 https://github.com/microsoft/vcpkg/pull/28656

The conclusion was to add new gz-port while maintain all the old other port with their version number.

File: ports/sdformat13/cmake-config.patch (hunk 1 above)
Comment (Adela0814) [inline ports/sdformat13/cmake-config.patch:9]:
```suggestion
+install(TARGETS using_parser_urdf EXPORT unofficial-${PROJECT_LIBRARY_TARGET_NAME} COMPONENT libraries)
```

**Review by Adela0814 — CHANGES_REQUESTED** [review]

File: ports/sdformat13/no-absolute.patch (hunk 3 above)
Comment (traversaro) [reply ports/sdformat13/no-absolute.patch:19]: sdformat looks into folders using an hardcoded version of `CMAKE_INSTALL_PREFIX` set at install time, that make sense for packages that are not moved after CMake install (like debian packages) but it does not make sense for vcpkg. The overall issue is tracked in https://github.com/gazebosim/gz-sim/issues/626, but I realized thanks to this comment that we dd not fix the issue for sdformat.

File: ports/sdformat13/cmake-config.patch (hunk 1 above)
Comment (talregev) [reply ports/sdformat13/cmake-config.patch]: Thank you! 
Done.

**Comment by talregev** [comment]
> > > The error is the one described in [#31280 (comment)](https://github.com/microsoft/vcpkg/pull/31280#issuecomment-1537483083) , nothing related to the use of `gz_` CMake function.
> > 
> > 
> > `using_parser_urdf` is an interface target created in sdformat13 `src/CMakeLists.txt`, and linked to the main library target. This is the standard CMake stuff. However, the export of CMake config is hidden in gz functions. This export would need to handle the interface target as well.
> 
> Ahh, sorry I misunderstdood, that's totally correct.
> 
> > @traversaro
> > Please create a PR in upstream, and I will make a patch from it.
> 
> I am happy to give you the hints to understand how to fix it. To be honest I am not interest at the moment in building sdformat13 as part of vcpkg and in static mode, so I prefer not to prepare the PR and submit it.
> 
> The `sdformat13` library is created in https://github.com/gazebosim/gz-cmake/blob/gz-cmake3_3.1.0/cmake/GzUtils.cmake#L1540 and installed in https://github.com/gazebosim/gz-cmake/blob/gz-cmake3_3.1.0/cmake/GzUtils.cmake#L1611 , so the `EXPORT` used for it is `sdformat13` . The target installed with that export are installed in https://github.com/gazebosim/gz-cmake/blob/88e7dc5f727e2270f206f19f558ab8f7fb5b8df1/cmake/GzPackaging.cmake#L372 . So a simple workaround is to just install the `using_parser_urdf` with the `sdformat13` export, close to where `using_parser_urdf` library is defined with `add_library`.

@traversaro Will you consider the next version of sdformat will support vcpkg in static mode?

**Comment by talregev** [comment]
@Adela0814 adding `unofficial-` as a prefix, create to build failure:
```
CMake Error: install(EXPORT "sdformat13" ...) includes target "sdformat13" which requires target "using_parser_urdf" that is not in any export set.
CMake Error in src/CMakeLists.txt:
  export called with target "sdformat13" which requires target
  "using_parser_urdf" that is not in any export set.
```

Please read comments here: 
https://github.com/microsoft/vcpkg/pull/31280#issuecomment-1546619308

I want to revert this change.

**Comment by traversaro** [comment]
> @traversaro Will you consider the next version of sdformat will support vcpkg in sperate static and dynamic mode?

I am not a mantainer of sdformat, but I guess that if you submit a PR based on https://github.com/microsoft/vcpkg/pull/31280#issuecomment-1546619308 explaining in which case it is useful I guess the mantainers would be happy to review it.

**Comment by talregev** [comment]
> > @traversaro Will you consider the next version of sdformat will support vcpkg in sperate static and dynamic mode?
> 
> I am not a mantainer of sdformat, but I guess that if you submit a PR based on [#31280 (comment)](https://github.com/microsoft/vcpkg/pull/31280#issuecomment-1546619308) explaining in which case it is useful I guess the mantainers would be happy to review it.

If I knew how to solve it, I was create a PR and a patch. I am lacking the knowledge how to do it.

**Comment by talregev** [comment]
> @Adela0814 adding `unofficial-` as a prefix, create to build failure:
> 
> ```
> CMake Error: install(EXPORT "sdformat13" ...) includes target "sdformat13" which requires target "using_parser_urdf" that is not in any export set.
> CMake Error in src/CMakeLists.txt:
>   export called with target "sdformat13" which requires target
>   "using_parser_urdf" that is not in any export set.
> ```
> 
> Please read comments here: [#31280 (comment)](https://github.com/microsoft/vcpkg/pull/31280#issuecomment-1546619308)
> 
> I want to revert this change.

@Adela0814 I will happy to hear your comment about the build fail.

**Comment by talregev** [comment]
@Adela0814 It now compile also on windows.
Please review it again.

**Review by Adela0814 — APPROVED** [review]

**Comment by talregev** [comment]
@Adela0814 Thank you for your approve.
I see this PR didn't get in the previous batch. 
Are we waiting for more review?

**Comment by talregev** [comment]
Thank you!

**Pull request merged**